### DISCUSSION TITLE
Rainsensor Optical Rain Gauge RG-15

### DISCUSSION THREAD

**Discussion by Mark-Muc** [arendst/Tasmota#22612] [Q&A]
Hello . I bought a Rainsensor Optical Rain Gauge RG-15
https://www.antratek.de/optical-rain-gauge-rg-15
I connected to ESP32 Wemos LOLIN32 V1.0.0. So far it's working fine.
compiled with:
```
#ifndef USE_HRG15
#define USE_HRG15
#endif
```

Connected more devices. looks like this
![image](https://github.com/user-attachments/assets/3cbc5e24-24d0-46e5-9cf9-9aa7cee3b99c)

My problem is: with MQTT Name ..
in the console the reporting looks like this:
MQTT:
```
"RG-15": {
    "Active": 0,
    "Event": 6.6,
    "Total": 64.19,
    "FlowRate": 0
  },
```
I tried to use the name "RG-15" on Node red .. read mqtt node  but it seems to be, that NODE RED don't like "-" in a MQTT Name, simple it crashes .. 
So I wrote a small rule:
```
Rule2
ON tele-RG-15#Active DO publish tele/%topic%/RG15/Active %value% ENDON
ON tele-RG-15#Event DO publish tele/%topic%/RG15/Event %value% ENDON
ON tele-RG-15#Total DO publish tele/%topic%/RG15/Total %value% ENDON
ON tele-RG-15#FlowRate DO publish tele/%topic%/RG15/FlowRate %value% ENDON

```
so I got in MQTT:
```
RG15
Active = 0.00
Total = 64.20
FlowRate = 0.02
Event = 6.61
```
but it seems that this use a lot of traffic. so sometimes the state of the Wemos LOLIN32 V1.0.0  is undefined and resets the relays.
No big deal so far, but if we could rename the name from now:
`RG-15`
to
`RG15`
it would help a lot ..
Someone have the same experience?

Regards 
Mark-Muc

**Comment by arendst** [comment]
Tx.

I'll change it now as I agree, the dash should never have been used in a JSON name unless it could have been changed by an underline as "normal" tasmota option.

This time I'll remove the dash as you suggest. Hold on.

**Reply by Mark-Muc** [reply]
oh thanky you . sure u can use "_" aswell but RG15 would be better!
THX
Mark-Muc

**Comment by Mark-Muc** [comment]
is the name already changed??
THX
Mark-Muc

**Comment by arendst** [comment]
Yep. Changed in latest dev

**Comment by Mark-Muc** [comment]
thank you . I'll have a look
Mark-Muc

**Comment by Mark-Muc** [comment]
Just to let people know!
Now everything is working fine and NO problems. Run it now for 24h und no relays have been switched.
Again Thank you!!
Mark-Muc

**Reply by Noschvie** [reply]
Hi @Mark-Muc
how long is the connection cable between ESP and the RG-15 and how did you solve the power supply, do you use the 5V or 3.3V from the ESP board? Thanks!

**Reply by Mark-Muc** [reply]
Hello Norbert!!
The sensor is outside my greenhouse . I do have a electrical distribution in the greenhouse .. so maybe it's about 70 cm to the ESP32.
I have in the electrical distribution a 12V 1,25V power supply an connected this to GND and V+ .. you also have to connect GND of the sensor to GND of the processor so I don't use the 5V / 3.3V power supply from the processor. 
Hope it helps
Mark-Muc
![image](https://github.com/user-attachments/assets/4f20415a-5336-4a84-8386-f8c4a1af0ac4)

**Reply by Noschvie** [reply]
A picture says more than many words...

**Comment by Noschvie** [comment]
Is there error handling if communication to the sensor is interrupted?

**Reply by Mark-Muc** [reply]
that's I'll really don't know, maybe u can ask the manufacture. It simpy say's NULL if no date come in or  disconnected.
In my situation the OUT channel don't work. The doc say's it should be low it any rain impulse come in, but it doesn't .. 
I run it on tasmota -> NODE-RED -> HA . so it's no problem to detect rain and built a own binary_sensor.
Hope it helps 
Mark-Muc

**Comment by Noschvie** [comment]
Have the sensor up and running.
If the serial connection is interrupted, there is no indication of this. I'll take a look at the code to see if I can log a connection error.

**Reply by Mark-Muc** [reply]
one short question: Does the` J1 OUT` channel work for u??
THX
Mark-Muc

**Reply by Noschvie** [reply]
Sorry, have the `J1 OUT` not connected, just `J2 GND, V+, RS232 IN, RS232 OUT`

**Comment by Noschvie** [comment]
The problem seems to be related to Hydreon Serial->available().
If the serial is interrupted during operation, this is not recognized correctly; HydreonSerial->available() remains unchanged at 0.

`HydreonSerial = new TasmotaSerial(Pin(GPIO_HRG15_RX), Pin(GPIO_HRG15_TX), 2, 0, RG15_BUFFER_SIZE);`

Should TasmotaSerial detect an interruption? A reconnect is carried out correctly.
A "rain event has timed out" is only recognized if the connection is interrupted during booting.

**Reply by sfromis** [reply]
How would you even expect `TasmotaSerial` to detect an "interruption"? How do you even define what that word should mean in terms of pins and what's going on at the physical layer?

The way async serial works is that each byte is sent separately (with start/stop/parity bits added), and in between bytes there is nothing going on, meaning that it makes no difference if you cut the cable.

Of course, you could talk about detecting lack of connection at the protocol level, having expectations about what bytes should arrive, but `TasmotaSerial` can't "know" that.

**Reply by Noschvie** [reply]
Ok, thanks for your explanation.
That's why we should improve the driver in order to be able to detect a timeout when receiving the sensor data.
I will try to trigger a timeout event if there is no sensor data for at least 60 seconds.
